src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+  arma::vec scores(2 * numProj);
+  scores.rows(0, numProj - 1) = arma::pow(limLow, 2);
+  scores.rows(numProj, 2 * numProj - 1) = arma::pow(limHigh, 2);
+
+  // actions vector shows what transformation to apply to a coordinate
+  arma::Col<short int> actions(2 * numProj); // will be [-1 ... 1 ...]
+
+  actions.rows(0, numProj - 1) = // first numProj rows
+    -1 * arma::ones< arma::Col<short int> > (numProj); // -1s
+
+  actions.rows(numProj, 2 * numProj - 1) = // last numProj rows
+    arma::ones< arma::Col<short int> > (numProj); // 1s
+
+
+  // acting dimension vector shows which coordinate to transform according to
+  // actions described by actions vector
+  arma::Col<size_t> positions(2 * numProj); // will be [0 1 2 ... 0 1 2 ...]
+  positions.rows(0, numProj - 1) =
+    arma::linspace< arma::Col<size_t> >(0, numProj - 1, numProj);
+  positions.rows(numProj, 2 * numProj - 1) =
+    arma::linspace< arma::Col<size_t> >(0, numProj - 1, numProj);
+
+  // optimization: no need to create heap for 1 or 2 codes
+  if (T <= 2)
+  {
+
+    // First, find location of minimum score, generate 1 perturbation vector,
+    // and add its code to additionalProbingBins column 0.
+
+    // find location and value of smallest element of scores vector
+    double minscore = scores[0];
+    size_t minloc = 0;
+    for (size_t s = 1; s < 2 * numProj; ++s)
+    {
+      if (minscore > scores[s])
+      {
+        minscore = scores[s];
+        minloc = s;
+      }
+    }
+    
+    // add or subtract 1 to dimension corresponding to minimum score
+    additionalProbingBins(positions[minloc], 0) += actions[minloc];
+    if (T == 1)
+      return; //done if asked for only 1 code
+
+    // Now, find location of second smallest score and generate one more vector.
+    // The second perturbation vector still can't comprise of more than one
+    // change in the bin codes. This is because of the way perturbation vectors
+    // are generated: First we create the one with the smallest score (Ao) and
+    // then we either add 1 extra dimension to it (Ae) or shift it by one (As).
+    // Since As contains the second smallest score, and Ae contains both the
+    // smallest and the second smallest, it's obvious that score(Ae) >
+    // score(As). Therefore the second perturbation vector is ALWAYS the vector
+    // containing only the second-lowest scoring perturbation.
+    
+    
+    double minscore2 = scores[0];
+    size_t minloc2 = 0;
+    for (size_t s = 0; s < 2 * numProj; ++s) // here we can't start from 1
+    {
+      if ( minscore2 > scores[s] && s != minloc) //second smallest
+      {
+        minscore2 = scores[s];
+        minloc2 = s;
+      }
+    }
+
+    // add or subtract 1 to create second-lowest scoring vector
+    additionalProbingBins(positions[minloc2], 1) += actions[minloc2];
+    return;
+    
+  }
+
+  // sort everything in increasing order
+  arma::Col<long long unsigned int> sortidx = arma::sort_index(scores);
```

Comment:
Why `long long unsigned int`?  Either `size_t` or possibly `arma::uword` (I prefer the former but sometimes it is not possible) should work here.